tests/test_browser.py

```diff
@@ -56,6 +56,8 @@ def shell_with_script(shell_file, output_file, replacement):
     with open(output_file, 'w') as output:
       output.write(input.read().replace('{{{ SCRIPT }}}', replacement))
 
+def requires_hardware(func):
+  return lambda self: self.skip("This test requires hardware access including graphics and sound devices") if os.environ.get("EM_LACKS_HARDWARE_ACCESS") else func(self)
```

Comment:
This could be a bit vague - perhaps we change this to `requires_gpu_hardware` and `requires_audio_playback_support` or similar, so that it tells which aspect is necessary. Reading a long list of `@requires_hardware` below could be understood as "this won't run in a VM but needs a nonvirtualized system"?

Reply:
Yeah... what about this?

```py
def requires_hardware(requires="graphics"):
  return lambda func: lambda self: self.skip("This test requires %s hardware access" % requires) if os.environ.get("EM_LACKS_HARDWARE_ACCESS") else func(self)
```

So that we can use as:

```py
@requires_hardware()
def test_requires_graphics:
  pass

@requires_hardware("audio")
def test_requires_audo:
  pass
```